server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -0,0 +1,241 @@
+                var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+                clusterService.submitStateUpdateTask("health-metadata-update", task, config, taskExecutor);
+            }
+        }
+    }
+
+    private void resetHealthMetadata(String source) {
+        var task = UpsertHealthMetadataTask.createInsertTask(settings);
+        var config = ClusterStateTaskConfig.build(Priority.NORMAL);
+        clusterService.submitStateUpdateTask(source, task, config, taskExecutor);
+    }
+
+    public static List<NamedXContentRegistry.Entry> getNamedXContentParsers() {
+        return List.of(
+            new NamedXContentRegistry.Entry(Metadata.Custom.class, new ParseField(HealthMetadata.TYPE), HealthMetadata::fromXContent)
+        );
+    }
+
+    public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        return List.of(
+            new NamedWriteableRegistry.Entry(Metadata.Custom.class, HealthMetadata.TYPE, HealthMetadata::new),
+            new NamedWriteableRegistry.Entry(NamedDiff.class, HealthMetadata.TYPE, HealthMetadata::readDiffFrom)
+        );
+    }
+
+    static class UpsertHealthMetadataTask implements ClusterStateTaskListener {
+
+        private final HealthMetadata healthMetadata;
+
+        private UpsertHealthMetadataTask(HealthMetadata healthMetadata) {
+            this.healthMetadata = healthMetadata;
+        }
+
+        static UpsertHealthMetadataTask createInsertTask(Settings settings) {
+            return new UpsertHealthMetadataTask(
+                new HealthMetadata(
+                    new HealthMetadata.Disk(
+                        HealthMetadata.Disk.Threshold.parse(
+                            CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING.get(settings),
+                            CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING.getKey()
+                        ),
+                        HealthMetadata.Disk.Threshold.parse(
+                            CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING.get(settings),
+                            CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING.getKey()
+                        ),
+                        HealthMetadata.Disk.Threshold.parse(
+                            CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.get(settings),
+                            CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.getKey()
+                        ),
+                        new HealthMetadata.Disk.Threshold(CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_FROZEN_SETTING.get(settings)),
+                        CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_FROZEN_MAX_HEADROOM_SETTING.get(settings)
+                    )
+                )
+            );
+        }
+
+        static UpsertHealthMetadataTask createUpdateTask(ClusterState clusterState, String setting, String value) {
```

Comment:
Shall we document this?

Reply:
While documenting this, I noticed that this had a bug. The code was applying the changes on the cluster state it from the caller. Since we are working with batched updates, the last update applied would override all the previous ones. 

Fixed and now the health metadata is constructed in the execute method which makes much more sense.

Reply:
++ good catch 

I've just added one more suggestion to have the `execute` method be `ClusterState -> ClusterState` (this would avoid an ignored parameter and having it @Nullable)